yjit/src/codegen.rs

```diff
@@ -5792,6 +5792,82 @@ fn jit_rb_str_byteslice(
     true
 }
 
+fn jit_rb_str_aref_m(
+    jit: &mut JITState,
+    asm: &mut Assembler,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<BlockHandler>,
+    argc: i32,
+    _known_recv_class: Option<VALUE>,
+) -> bool {
+    // In yjit-bench the most common usages by far are single fixnum or two fixnums.
+    // rb_str_substr should be leaf if indexes are fixnums
+    if argc == 2 {
+        match (asm.ctx.get_opnd_type(StackOpnd(0)), asm.ctx.get_opnd_type(StackOpnd(1))) {
+            (Type::Fixnum, Type::Fixnum) => {},
+            // There is a two-argument form of (RegExp, Fixnum) which needs a different c func.
+            // Other types will raise.
+            _ => { return false },
+        }
+    } else if argc == 1 {
+        match asm.ctx.get_opnd_type(StackOpnd(0)) {
+            Type::Fixnum => {},
+            // Besides Fixnum this could also be a Range or a RegExp which are handled by separate c funcs.
+            // Other types will raise.
+            _ => {
+                // If the context doesn't have the type info we try a little harder.
+                let comptime_arg = jit.peek_at_stack(&asm.ctx, 0);
+                let arg0 = asm.stack_opnd(0);
+                if comptime_arg.fixnum_p() {
+                    asm.test(arg0, Opnd::UImm(RUBY_FIXNUM_FLAG as u64));
+
+                    jit_chain_guard(
+                        JCC_JZ,
+                        jit,
+                        asm,
+                        SEND_MAX_DEPTH,
+                        Counter::guard_send_not_fixnums,
```

Comment:
We overloaded it while pairing, but let's add a separate counter for this. Something like `gurad_send_str_aref_not_fixnum`?

Reply:
Sounds good.  I added a line for it to `stats.rs`, is that all that is needed?

Reply:
You also need to print the value in `yjit.rb`.

We should probably have some automated/default formatting there 😅

Reply:
I think it would be covered by this one:
https://github.com/ruby/ruby/blob/ee1cd1656fc667840282a4e9c3f2cc5e3154e50e/yjit.rb#L293
but I'm not sure what code would prove it.

Reply:
```
method call exit reasons:
    str_aref_not_fixnum:          6 (100.0%)
```